### PR TITLE
#824 getters and setters

### LINKED ISSUE #824: [IC][Schema][Programs] Getters/setters for new schema fields and tables
Description:
#819 introduces several new schema tables and fields. #815 in particular would utilize getters and setters well.

Comments:

`MasterList`
- `getIsProgramForStore` - checking if the master list is a program and usable by the current store
- `getOrderTypes` - gets the order types for the program for the current store

... possibly more

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #824 

### NOTE: Branches from #819 

Adds additional sync logic, schema update and setters/getters for new IC tables.

### CODE REVIEW

File: src/database/DataTypes/PeriodSchedule.js [hunk 1]
```diff
@@ -5,7 +5,24 @@
...
+    const periods = this.periods.reduce(period => {
+      if (period.numberOfRequisitionsForProgram(program) >= maxMOS) return null;
+      return period;
+    });
+    return periods;
+  }
+
+  addPeriod(period) {
+    this.periods.push(period);
+  }
```
Comment (Chris-Petty) [inline src/database/DataTypes/PeriodSchedule.js]: Just a thought: this method is a one liner and likely will only be used in `addPeriodIfUnique`, which we'll usually (always?) want people to use. Just inline it in addPeriodIfUnique?

File: src/sync/incomingSyncUtils.js [hunk 2]
```diff
@@ -552,7 +555,7 @@ export const createOrUpdateRecord = (database, settings, recordType, record) =>
     }
     case 'Store': {
       if (settings.get(THIS_STORE_ID) === record.ID) {
-        settings.set(THIS_STORE_ID, record.StoreID);
+        settings.set(THIS_STORE_TAGS, record.tags);
```
Comment (Chris-Petty) [inline src/sync/incomingSyncUtils.js]: Good catch haha

File: src/database/DataTypes/MasterList.js [hunk 3]
```diff
@@ -18,6 +18,35 @@ export class MasterList extends Realm.Object {
     if (this.items.filtered('id == $0', masterListItem.id).length > 0) return;
     this.addItem(masterListItem);
   }
+
+  getStoreTagObject(tags) {
+    const storeTags = tags.split(/[\s,]+/);
+    return Object.keys(JSON.parse(this.programSettings)).filter(
+      programsStoreTags => storeTags.indexOf(programsStoreTags) >= 0
```
Comment (andreievg) [inline src/database/DataTypes/MasterList.js]: should be iterating through this.programSettings.storeTags

File: src/database/DataTypes/MasterList.js [hunk 4]
```diff
@@ -18,6 +18,35 @@ export class MasterList extends Realm.Object {
     if (this.items.filtered('id == $0', masterListItem.id).length > 0) return;
     this.addItem(masterListItem);
   }
+
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/database/DataTypes/MasterList.js]: 100% - I was hoping I could get away with it and make an issue for next week. I'll do them shortly

File: src/database/DataTypes/Period.js [hunk 5]
```diff
@@ -5,7 +5,24 @@
...
+    return this.requisitions.filtered('program.id = $0', program.id).length;
+  }
+
+  addRequisition(requisition) {
+    this.requisitions.push(requisition);
+  }
+
+  addRequisitionIfUnique(requisition) {
+    if (this.requisitions.filtered('id == $0', requisition.id).length > 0) return;
+    this.addBatch(requisition);
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/database/DataTypes/Period.js]: yeah ta!

File: src/database/DataTypes/Period.js [hunk 6]
```diff
@@ -18,5 +35,6 @@ Period.schema = {
     endDate: { type: 'date', default: new Date() },
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/database/DataTypes/Period.js]: shouldn't be a problem, just defaults for ordering in a sync pull

File: src/database/DataTypes/PeriodSchedule.js [hunk 7]
```diff
@@ -5,7 +5,24 @@
 
 import Realm from 'realm';
 
-export class PeriodSchedule extends Realm.Object {}
+export class PeriodSchedule extends Realm.Object {
+  getUseablePeriodsForProgram(program, maxMOS) {
+    const periods = this.periods.reduce(period => {
+      if (period.numberOfRequisitionsForProgram(program) >= maxMOS) return null;
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/database/DataTypes/PeriodSchedule.js]: nah i'm dum dum

File: src/database/DataTypes/PeriodSchedule.js (hunk 1 above)
Comment (joshxg) [reply src/database/DataTypes/PeriodSchedule.js]: cool, was following suit of the other models, removed

File: src/sync/incomingSyncUtils.js (hunk 2 above)
Comment (joshxg) [reply src/sync/incomingSyncUtils.js]: oopsies

File: src/database/DataTypes/MasterList.js (hunk 3 above)
Comment (joshxg) [reply src/database/DataTypes/MasterList.js]: thanks

File: src/database/DataTypes/MasterList.js [hunk 8]
```diff
@@ -18,6 +18,43 @@ export class MasterList extends Realm.Object {
...
+   * Find if this master list is a program useable by the current store
+   * @param  {string}  tags   Current stores tags field
+   * @return {bool} true if the current store can use this master list
+   */
+  canUseProgram(tags) {
+    return !!this.getStoreTagObject(tags);
+  }
+
+  /**
+   * Find a specificly named order type in the store tag object for this store and masterlist
```
Comment (Chris-Petty) [inline src/database/DataTypes/MasterList.js]:
```suggestion
   * Find a specifically named order type in the store tag object for this store and masterlist
```

**Comment by joshxg** [comment]
fixed

**Pull request merged**